In [72]:
'''Problem Statement 1:
Is gender independent of education level? A random sample of 395 people were 
surveyed and each person was asked to report the highest education level they 
obtained. The data that resulted from the survey is summarized in the following table:
       High School  Bachelors  Masters  Ph.d.  Total
Female     60         54          46    41     201
Male       40         44          53    57     194
Total     100         98          99    98     395
Question: Are gender and education level dependent at 5% level of significance? In
other words, given the data collected above, is there a relationship between the
gender of an individual and the level of education that they have obtained? '''


# first we to make dataframe same as the table given : 

f = [60,54,46,41,201]
m = [40,44,53,57,194]
t = [100,98,99,98,395]
import pandas as pd
import numpy as np
import scipy.stats as stats

data = []
data.append(f)
data.append(m)
data.append(t)

df = pd.DataFrame(data)# it will add like rows 

df.columns = ['high School', 'becholers', 'masters','P.hd','row_total']
df.index = ['females', 'males', 'col_total']
df


# null hypothesis is - there is no relationship between gender & qualifications

df

high School  becholers  masters  P.hd  row_total
females             60         54       46    41        201
males               40         44       53    57        194
col_total          100         98       99    98        395

In [73]:
# here we have to check chi square test in two dimensions but here we have to calculate the expected count corroesponding to
# each value which we can get by multiplying that row total with columns total & divide it by full total. in short we can use 
# np.outer function & pass one parameter as row total & 2nd as columns total, then divide it by whole total

exp_array = np.outer(df['row_total'][0:2], df.loc['col_total'][0:4]) / 395

exp_df = pd.DataFrame(df1)
exp_df.columns = ['high School', 'becholers', 'masters','P.hd']
exp_df.index = ['females', 'males']
exp_df

high School  becholers    masters       P.hd
females    50.886076  49.868354  50.377215  49.868354
males      49.113924  48.131646  48.622785  48.131646

In [74]:
# hence we get the expected values of all the counts, now we calculate chi square statistics, for that we have to 
# create another dataframe from original one &  removing the col_total & row_total, so that its dimensions be same like exp_df
obs_df = df.iloc[0:2,0:4]
obs_df

high School  becholers  masters  P.hd
females           60         54       46    41
males             40         44       53    57

In [75]:
# We call .sum() twice:once to get the column sums and a second time to add the column sums together,
# returning the sum of the entire 2D table
chi_stat = ((obs_df - exp_df)**2 / exp_df).sum().sum()
chi_stat

8.006066246262538

so we find out that chi square value which is 8

In [77]:
# now our table is 2X4, & the degree of freedom is equals to the product of row & column minus one from both of them 
# so our dof is : 1 X 3 = 3
# now we found critical value with alpha = 5% or C.I. 95% & then p value

critical_value = stats.chi2.ppf(q= 0.95, df = 3)

print('critical value : {}'.format(critical_value))

p_value = 1 - stats.chi2.cdf(x = chi_stat, df = 3)

print('p value is : {}'.format(p_value))

critical value : 7.814727903251179
p value is : 0.04588650089174717


Use stats.chi2_contingency() function to conduct a test of independence automatically given a frequency table of observed counts:

In [78]:
stats.chi2_contingency(observed = obs_df)

(8.006066246262538,
 0.045886500891747214,
 3,
 array([[50.88607595, 49.86835443, 50.37721519, 49.86835443],
        [49.11392405, 48.13164557, 48.62278481, 48.13164557]]))

so it shows that chi square value is 8, p value is 0.045 & then the array having expected counts. now the critical value with 
3df is 7.815 which is smaller than chi_sq value, so we reject the null hypothesis and we can say that there is some relation
in between the gender & their education qualifications with significance level 5%

In [3]:
'''Problem Statement 2:
Using the following data, perform a oneway analysis of variance using α=.05. Write up the results in APA format.

[Group1: 51, 45, 33, 45, 67] [Group2: 23, 43, 23, 43, 45] [Group3: 56, 76, 74, 87, 56] '''

# scipy labrary has a function f_oneway() to test one way anova test
import scipy.stats as stats

Group1= [51, 45, 33, 45, 67]
Group2= [23, 43, 23, 43, 45] 
Group3= [56, 76, 74, 87, 56]

statistics , pvalue = stats.f_oneway(Group1,Group2, Group3)

print('the f static value is : {} & p value is : {}'.format(statistics, pvalue))

if p_value < 0.05:
    print('True')
else:
    print('False')
    

the f static value is : 9.747205503009463 & p value is : 0.0030597541434430556


In [5]:
'''Problem Statement 3:
Calculate F Test for given 10, 20, 30, 40, 50 and 5,10,15, 20, 25.
For 10, 20, 30, 40, 50: '''

Group1 = [10,20,30,40,50]
Group2 = [5,10,15,20,25]

import numpy as np
mean_1 = np.mean(Group1)
mean_2 = np.mean(Group2)

grp1_sub_mean1 = []
grp2_sub_mean2 = []
add1 = 0
add2 = 0
for items in Group1:
    add1 += (items - mean_1)**2
for items in Group2:
    add2 += (items - mean_2)**2
var1 = add1/(len(Group1)-1)
var2 = add2/(len(Group2)-1)

F_Test = var1/var2
print("F Test for given 10, 20, 30, 40, 50 and 5,10,15, 20, 25 is : ",F_Test)


F Test for given 10, 20, 30, 40, 50 and 5,10,15, 20, 25 is :  4.0
